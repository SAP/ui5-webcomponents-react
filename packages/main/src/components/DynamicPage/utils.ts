export const DynamicPageCssVariables = {
  headerDisplay: '--_ui5wcr_DynamicPage_header_display',
  titleFontSize: '--_ui5wcr_DynamicPage_title_fontsize'
};
