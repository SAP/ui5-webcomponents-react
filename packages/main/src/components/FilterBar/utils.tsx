import { cloneElement } from 'react';

const inputTagNames = new Set([
  'UI5-COMBOBOX',
  'UI5-DATE-PICKER',
  'UI5-DATERANGE-PICKER',
  'UI5-DATETIME-PICKER',
  'UI5-DURATION-PICKER',
  'UI5-INPUT',
  'UI5-MULTI-INPUT',
  'UI5-RATING-INDICATOR',
  'UI5-STEP-INPUT',
  'UI5-TIME-PICKER'
]);

export const filterValue = (ref, child) => {
  const tagName = ref.tagName;
  let filterItemProps = {};
  if (inputTagNames.has(tagName)) {
    filterItemProps = { value: ref.value };
  }
  if (tagName === 'UI5-SELECT' || tagName === 'UI5-MULTI-COMBOBOX') {
    const selectedIndices = Array.from(ref.children as HTMLCollectionOf<any>)
      .map((item, index) => (item.selected ? index : false))
      .filter((el) => el !== false);
    const selectedIndicesSet = new Set(selectedIndices);
    const options = child.props.children.props.children?.map((item, index) => {
      if (selectedIndicesSet.has(index)) {
        return cloneElement(item, { selected: true });
      }
      return cloneElement(item, { selected: false });
    });
    filterItemProps = { children: options };
  }
  if (tagName === 'UI5-SWITCH' || tagName === 'UI5-CHECKBOX') {
    filterItemProps = { checked: ref.checked };
  }
  return filterItemProps;
};

export const renderSearchWithValue = (renderSearchElement, searchValue) => {
  return cloneElement(renderSearchElement, { value: searchValue ?? renderSearchElement?.props?.value });
};
