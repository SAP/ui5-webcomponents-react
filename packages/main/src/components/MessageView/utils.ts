import iconAlert from '@ui5/webcomponents-icons/dist/alert.js';
import iconError from '@ui5/webcomponents-icons/dist/error.js';
import iconInformation from '@ui5/webcomponents-icons/dist/information.js';
import iconSysEnter from '@ui5/webcomponents-icons/dist/sys-enter-2.js';
import { ValueState } from '../../enums/ValueState';

export const getIconNameForType = (type: ValueState | keyof typeof ValueState): string => {
  switch (type) {
    case ValueState.Error:
    case 'Error':
      return iconError;
    case ValueState.Success:
    case 'Success':
      return iconSysEnter;
    case ValueState.Warning:
    case 'Warning':
      return iconAlert;
    default:
      return iconInformation;
  }
};
