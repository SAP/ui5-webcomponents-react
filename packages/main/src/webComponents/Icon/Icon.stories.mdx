import { ArgsTable, Canvas, Meta, Story } from '@storybook/addon-docs';
import { Icon } from './index';
import '@ui5/webcomponents-icons/dist/employee.js';
import { DocsHeader } from '@docs/DocsHeader';
import { Footer } from '@docs/Footer';

<Meta
  title="Data Display / Icon"
  component={Icon}
  args={{
    name: 'employee',
    style: {},
    className: '',
    tooltip: '',
    slot: '',
    ref: null
  }}
/>

<DocsHeader />

<br />

## Example

<Canvas>
  <Story name="Default">
    {(args) => {
      return <Icon {...args} />;
    }}
  </Story>
</Canvas>

## Properties

<ArgsTable story="." />

<Footer />
