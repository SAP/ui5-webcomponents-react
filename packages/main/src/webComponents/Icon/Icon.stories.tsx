// @ts-expect-error: storybook can load json
import cem from '@sb/custom-element-manifests/main.json';
import type { Meta, StoryObj } from '@storybook/react';
import employeeIcon from '@ui5/webcomponents-icons/dist/employee.js';
import { Icon } from './index.js';

//todo: remove custom description once JSDoc transforming is fixed
const description = cem.modules
  .find((item) => item.path === 'dist/Icon.js')
  .declarations.find((item) => item.kind === 'class').description;

function iconSummaryFormatter(htmlDesc: string) {
  const summary = htmlDesc.replace(/###\s?Overview\n*/, '').replace(/### ES6 Module Import\n*`.+`/, '');
  return summary.replaceAll('ui5-icon', 'Icon');
}

const meta = {
  title: 'Data Display / Icon',
  component: Icon,
  argTypes: {},
  args: {
    name: employeeIcon
  },
  parameters: {
    docs: {
      description: {
        component:
          iconSummaryFormatter(description) +
          '\n__Note:__ This is a UI5 Web Component! [Repository](https://github.com/SAP/ui5-webcomponents) | [Documentation](https://sap.github.io/ui5-webcomponents/playground/)'
      }
    }
  },
  tags: ['package:@ui5/webcomponents']
} satisfies Meta<typeof Icon>;

export default meta;
type Story = StoryObj<typeof meta>;

export const Default: Story = {};
