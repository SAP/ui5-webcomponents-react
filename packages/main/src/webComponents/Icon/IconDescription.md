## Usage

1\. **Get familiar with the icons collections.**  
Before displaying an icon, you need to explore the icons collections to find and import the desired icon.  
Currently there are 3 icons collection, available as 3 npm packages:  

*   <ui5-link target="_blank" href="https://www.npmjs.com/package/@ui5/webcomponents-icons" class="api-table-content-cell-link">@ui5/webcomponents-icons</ui5-link> represents the "SAP-icons" collection and includes the following <ui5-link target="_blank" href="https://openui5.hana.ondemand.com/test-resources/sap/m/demokit/iconExplorer/webapp/index.html#/overview/SAP-icons" class="api-table-content-cell-link">icons</ui5-link>.
*   <ui5-link target="_blank" href="https://www.npmjs.com/package/@ui5/webcomponents-icons-tnt" class="api-table-content-cell-link">@ui5/webcomponents-icons-tnt</ui5-link> represents the "tnt" collection and includes the following <ui5-link target="_blank" href="https://openui5.hana.ondemand.com/test-resources/sap/m/demokit/iconExplorer/webapp/index.html#/overview/SAP-icons-TNT" class="api-table-content-cell-link">icons</ui5-link>.
*   <ui5-link target="_blank" href="https://www.npmjs.com/package/@ui5/webcomponents-icons-business-suite" class="api-table-content-cell-link">@ui5/webcomponents-icons-icons-business-suite</ui5-link> represents the "business-suite" collection and includes the following <ui5-link target="_blank" href="https://sapui5.hana.ondemand.com/test-resources/sap/m/demokit/iconExplorer/webapp/index.html#/overview/BusinessSuiteInAppSymbols" class="api-table-content-cell-link">icons</ui5-link>.

2\. **After exploring the icons collections, add one or more of the packages as dependencies to your project.**  
`npm i @ui5/webcomponents-icons`  
`npm i @ui5/webcomponents-icons-tnt`  
`npm i @ui5/webcomponents-icons-business-suite`  
  
3\. **Then, import the desired icon**.  
`import "@ui5/{package_name}/dist/{icon_name}.js";`  
  
**For Example**:  
For the standard "SAP-icons" icon collection, import an icon from the `@ui5/webcomponents-icons` package:  
`import "@ui5/webcomponents-icons/dist/employee.js";`  
  
For the "tnt" (SAP Fiori Tools) icon collection, import an icon from the `@ui5/webcomponents-icons-tnt` package:  
`import "@ui5/webcomponents-icons-tnt/dist/antenna.js";`  
  
For the "business-suite" (SAP Business Suite) icon collection, import an icon from the `@ui5/webcomponents-icons-business-suite` package:  
`import "@ui5/webcomponents-icons-business-suite/dist/ab-testing.js";`  
  
4\. **Display the icon using the `Icon` web component.**  
Set the icon collection ("SAP-icons", "tnt" or "business-suite" - "SAP-icons" is the default icon collection and can be skipped)  
and the icon name to the `name` property.  
  
`<Icon name="employee"></Icon>`  
`<Icon name="tnt/antenna"></Icon>`  
`<Icon name="business-suite/ab-testing"></Icon>`

## Keyboard Handling

*   \[SPACE, ENTER, RETURN\] - Fires the `click` event if the `interactive` property is set to true.
*   \[SHIFT\] - If \[SPACE\] or \[ENTER\],\[RETURN\] is pressed, pressing \[SHIFT\] releases the Icon without triggering the click event.