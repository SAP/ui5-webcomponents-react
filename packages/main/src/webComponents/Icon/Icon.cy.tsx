import addIcon from '@ui5/webcomponents-icons/dist/add.js';
import { cypressPassThroughTestsFactory } from '@/cypress/support/utils';
import { Icon } from './index';

describe('Icon', () => {
  it('Basic Test (generated)', () => {
    cy.mount(<Icon name={addIcon} />);
    cy.get('[ui5-icon]').should('exist');
  });

  cypressPassThroughTestsFactory(Icon);
});
