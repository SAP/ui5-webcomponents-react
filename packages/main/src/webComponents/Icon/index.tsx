import { withWebComponent } from '@ui5/webcomponents-react/dist/withWebComponent';
import { CommonProps } from '@ui5/webcomponents-react/interfaces/CommonProps';

import '@ui5/webcomponents/dist/Icon.js';

export interface IconPropTypes extends CommonProps {
  /**
   * Defines the text alternative of the component. If not provided a default text alternative will be set, if present.
   *
   * **Note:** Every icon should have a text alternative in order to calculate its accessible name.
   */
  accessibleName?: string;
  /**
   * Defines if the icon is interactive (focusable and pressable)
   */
  interactive?: boolean;
  /**
   * Defines the unique identifier (icon name) of the component.
   * To browse all available icons, see the <ui5-link target="_blank" href="https://openui5.hana.ondemand.com/test-resources/sap/m/demokit/iconExplorer/webapp/index.html" class="api-table-content-cell-link">Icon Explorer</ui5-link>.
   * Example:
   * `name='add'`, `name='delete'`, `name='employee'`.
   *
   * **Note:** To use the SAP Fiori Tools icons, you need to set the `tnt` prefix in front of the icon's name.
   * Example:
   * `name='tnt/antenna'`, `name='tnt/actor'`, `name='tnt/api'`.
   */
  name?: string;
  /**
   * Defines whether the component should have a tooltip.
   */
  showTooltip?: boolean;
}

/**
 * The `Icon` component represents an SVG icon. There are two main scenarios how the `Icon` component is used: as a purely decorative element; or as a visually appealing clickable area in the form of an icon button.
 *
 * A large set of built-in icons is available and they can be used by setting the `name` property on the `Icon`. But before using an icon, you need to import the desired icon.
 * For the standard icon collection, you have to import an icon from the `@ui5/webcomponents-icons` package:
 * `import "@ui5/webcomponents-icons/dist/employee.js";`
 * For the SAP Fiori Tools icon collection (supported since 1.0.0-rc.10), you have to import an icon from the `@ui5/webcomponents-icons-tnt` package:
 * `import "@ui5/webcomponents-icons-tnt/dist/antenna.js";`
 *
 * <ui5-link href="https://sap.github.io/ui5-webcomponents/playground/components/Icon" target="_blank">UI5 Web Components Playground</ui5-link>
 */
const Icon = withWebComponent<IconPropTypes>(
  'ui5-icon',
  ['accessibleName', 'name'],
  ['interactive', 'showTooltip'],
  [],
  []
);

Icon.displayName = 'Icon';

export { Icon };
