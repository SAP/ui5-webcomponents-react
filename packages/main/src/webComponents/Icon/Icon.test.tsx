import { render } from '@shared/tests';
import '@ui5/webcomponents-icons/dist/add.js';
import { createCustomPropsTest, createWaitForDefineTest } from '@shared/tests/utils';
import { Icon } from '../Icon';
import React from 'react';

describe('Icon', () => {
  test('Basic Test (generated)', () => {
    const { asFragment } = render(<Icon name="add" />);
    expect(asFragment()).toMatchSnapshot();
  });
  createCustomPropsTest(Icon);
  createWaitForDefineTest(Icon);
});
