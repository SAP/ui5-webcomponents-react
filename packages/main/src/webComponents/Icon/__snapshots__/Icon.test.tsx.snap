// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Icon Basic Test (generated) 1`] = `
<DocumentFragment>
  <ui5-icon
    design="Default"
    name="add"
    ui5-icon=""
  />
</DocumentFragment>
`;
